src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,551 @@
+# coding=utf-8
+# Copyright 2018 DPR Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch DPR model for Open Domain Question Answering."""
+
+
+import collections
+import logging
+from typing import List, Optional, Tuple
+
+import torch
+from torch import Tensor as T
+from torch import nn
+from torch.serialization import default_restore_location
+
+from .configuration_dpr import DprConfig
+from .file_utils import add_start_docstrings
+from .modeling_bert import BertConfig, BertModel
+from .modeling_utils import PreTrainedModel
+
+
+logger = logging.getLogger(__name__)
+
+DPR_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "dpr-base-uncased",
+]
+
+
+#############
+# BertEncoder
+#############
+
+
+class DprBertEncoder(BertModel):
+    def __init__(self, config, project_dim: int = 0):
+        BertModel.__init__(self, config)
+        assert config.hidden_size > 0, "Encoder hidden_size can't be zero"
+        self.encode_proj = nn.Linear(config.hidden_size, project_dim) if project_dim != 0 else None
```

Comment:
```python 
if project != 0:
    self.encode_proj = ...
```

and then later check if the layer exists. This is also better for the logic regarding loading weights in the state_dict IMO and will make it easier to transform the model weights to TF model weights later.

Reply:
I'm not a big fan of having attributes may not exist depending of the config (because we would need to use `hasattr` and I think it's cleaner in python to just check is it is None).
I don't think this is a problem for loading weights either. It actually makes it easier if we want to load weights from the official DPR repo.

On the other hand I'm not familiar with the conversion to TF model weights. Is it blocking to have this argument that could be None ?

Reply:
Right now in the library we're usually doing what Patrick mentionned. You wouldn't have to use `hasattr` to check, but rather assign `self.project_dim = project_dim` and in the forward check `if self.project_dim != 0`

Reply:
Ok I see ! I updated it